addons/account/models/partner.py

```diff
@@ -51,6 +52,19 @@ def _check_zip(self):
             if position.zip_from and position.zip_to and position.zip_from > position.zip_to:
                 raise ValidationError(_('Invalid "Zip Range", please configure it properly.'))
 
+    @api.constrains('country_id', 'state_ids', 'foreign_vat')
+    def _validate_foreign_vat_country(self):
+        for record in self:
+            if record.foreign_vat and record.country_id == record.company_id.account_fiscal_country_id:
+                if record.foreign_vat == record.company_id.vat:
+                    raise ValidationError(_("You cannot create a fiscal position within your fiscal country with the same VAT number as the main one set on your company."))
+
+                if not record.state_ids:
+                    if record.company_id.account_fiscal_country_id.state_ids:
+                        raise ValidationError(_("You cannot create a fiscal position with a foreign VAT within your fiscal country without assigning it a state."))
+                    else:
+                        raise ValidationError(_("You cannot create a fiscal position with a foreign VAT within your fiscal country."))
```

Comment:
Why does it depend on the state? This is probably valid in the US, but I don't think it is the case worldwide.
This constraint would force you to create a fiscal position for all states of a country even if the vat number is the same

Or maybe this is only for that case: countries with multiple states and different VAT numbers?

Reply:
The field state_ids on fiscal positions is m2m; so if you need a unique foreign VAT available for a set of states, you should create only one fiscal positions, with all the right states listed ont it. 

Indeed, the only legit use case for foreign VAT within the same country  fiscal positions is to have different VAT numbers on them. This is the case in India, for example. These checks ensure that.